### ISSUE TITLE
Errors should not be adhoc in core

### ISSUE THREAD

**Issue by alabamenhu** [Raku/problem-solving#374]
Errors should not be `X::Adhoc` in core.

Adhoc errors only inform their issues via a message, which may not be easily parseable.  As well, per the documentation

> Note that young code will often be prototyped using X::AdHoc and then later be revised to use more specific subtypes of Exception. As such it is usually best not to explicitly rely on receiving an X::AdHoc – in many cases using the string returned by the .message method, which all Exceptions must have, is preferable.

However, this is complicated by the fact that the string returned by `.message` is also potentially unreliable, as LTA errors may be changed, improved, with data points in the message potentially being reorded, adjusted, etc.

**Comment by alabamenhu** [comment]
This is fairly simple (if laborious) thing to do.  The main issue is determining what error classes are needed so we're not creating a bunch of one-off errors when we could combine them.  Additionally, the `Exception`s created (and extant ones) should be updated to ensure that, in contrast to the documentation, any actionable information is contained in attributes which the `message` function uses to convert into plain English.  This will also make handling a potential internationalization of error messages much easier.

(I posted this in problem-solving because it's sort of a partially Raku, partially Rakudo issue.  I'm happy to go start working on a large PR updating all the `Exception`, but would want consensus on how to go about it first)

**Comment by vrurg** [comment]
I agree that it's both Raku and Rakudo issue at the same time. Changed the categorization of this problem accordingly.

Solving this problem would involve not only updating Rakudo sources but adding tests to the roast. This is what's on the surface.

Deeper goes the task of collecting the errors thrown using `nqp::die` throughout the sources of Rakudo and categorizing them.

Another task is to see what could be done about the errors coming from NQP and, worse even, from VMs. These are adhocs by definition as currently there is no mechanism of mapping them into HLL classes.

**Comment by alabamenhu** [comment]
Per something I noted in IRC, I think in updating the `Exception` base class, we should do something akin to 

    class Exception {
        method message { self.MESSAGE } 
        method MESSAGE is implementation-detail { 'Exception of type ' ~ self.WHAT.^name } 
    }

Each exception class would generate its message based on `MESSAGE` (name for an example only).  This would enable an internationalization module to do `INIT Exception.^find_method('message').wrap()` and not need to wrap the dozens of exception classes' methods which is not performant.

At the same time, in reworking them, I'd make a rule that all pertinent information goes in an attribution that makes logical sense to be used in a `grep`/`when` area.  The message just joins them into a human readable version.   A mostly good example of this is `X::Method::NotFound`, where the message *No such method 'foo' for invocant of type 'Bar'* is derived from the attributes `method` (=*foo*), `typename` (=*Bar*), and `private` (=*false*). It does however have a new one `addendum` that is basically an adhoc and could warrant some review in an `Exception` rework.

**Comment by vrurg** [comment]
> This would enable an internationalization module to do `INIT Exception.^find_method('message').wrap()`

I already dislike it:

1. `wrap` is a hack
2. one would have to do it for nearly every exception class with `message` and would have to follow any changes in the core
3. it would be necessary to know the internal implementation of complex `message` methods where a bunch of other methods are involved – as in your example with `X::Method::NoFound`

Besides, what worries me is that composite messages may require different order of their parts in other languages.

I think internationalization must be done in a totally different way at Raku core level. 

First of all, there must be some pluggable mechanism where forming of complex messages can be offloaded to external modules. For example, the core could provide defaults via `X::L10N::<LN>` modules, where `<LN>` is a language code.

Second, there must be a way of obtaining atomic translations for standard strings and it must work at VM level. At the Raku level it would then make sense to use `printf` instead of string concatenations to allow language-specific word ordering.

**Comment by alabamenhu** [comment]
> > This would enable an internationalization module to do `INIT Exception.^find_method('message').wrap()`
> 
> I already dislike it:
> 
>     1. `wrap` is a hack
> 
>     2. one would have to do it for nearly every exception class with `message` and would have to follow any changes in the core
> 
>     3. it would be necessary to know the internal implementation of complex `message` methods where a bunch of other methods are involved – as in your example with `X::Method::NoFound`
> 
> 
> Besides, what worries me is that composite messages may require different order of their parts in other languages.
> 
> I think internationalization must be done in a totally different way at Raku core level.
> 
> First of all, there must be some pluggable mechanism where forming of complex messages can be offloaded to external modules. For example, the core could provide defaults via `X::L10N::<LN>` modules, where `<LN>` is a language code.
> 
> Second, there must be a way of obtaining atomic translations for standard strings and it must work at VM level. At the Raku level it would then make sense to use `printf` instead of string concatenations to allow language-specific word ordering.

I just wonder if internationalized error messages should be something core handles.  The idea of having a single `message` that farms out to individual `MESSAGE`s means only one wrap needs to be handled and has minimal extra overhead but a much easier implementation, but also pretty much says empatically "I'm not a part of core".  If stuff is farmed out to a module, then complex messages can be handle pretty easily — frameworks are far more advanced today than the old `.po/.mo` standard. Do we consider the `Pod::To::` mechanism to be a part of Raku or Rakudo?  I think that's a good model to follow.  But that's probably a different problem solving issue (although certainly related — discussion of it sparked this issue).  

That said, the idea of including each piece of information (method, type, etc for `X::Method::NotFound`) would make rewriting complex messages much easier for such things.  But the primary goal would be to allow better handling of errors.  It's much easier to use `if .method eq 'foo'` or `if .type eq 'Bar'` on `X::Method::NotFound` than having to tease out that information based on the contents of `.message`.  (I use `X::Method::NotFound` here as it's a good example of how to do things, but also of where work needs to be done — the `addendum`).  If that's figure out, I think a lot more can be enabled.

**Comment by vrurg** [comment]
> I just wonder if internationalized error messages should be something core handles.

I think it should. It's not ought to do it all itself, but providing a pluggable API would be very beneficial.

Besides, having the API would also allow third party code to use it for their needs too.

And don't forget about VM where having a way of providing localized messages is needed too.

> means only one wrap needs to be handled

Any `wrap` is rather a hack. It is OK and sometimes even really useful when it comes to well isolated, controllable cases. But when it is supposed to affect the entire ecosystem of a program it is barely acceptable.

> but also pretty much says empatically "I'm not a part of core".

Does it really matter? Say, my typical `$LANG` is en_US.UTF-8, but if I set it to, say, uk_UA.UTF-8 and there would be no core support for it then installing something like `X::L10N::UA` (though, perhaps, more sensible name would be `uk_UA`) from the ecosystem makes it no less clear that it is not a core thing. But why should anybody care? As long as it doesn't break one's code. Either core or installed L10N modules would be auto-loaded by the core based on the locale making the process transparent to a user.

> frameworks are far more advanced today than the old `.po/.mo` standard.

I don't propose the exact replica of .po/.mo. Pluggability can mean that roles or classes in a L10N module may override methods of the original exceptions. In a way this is a cleaner replacement for `wrap`.
